src/components/Layout/GridTableLayout/GridTableLayout.stories.tsx

```diff
@@ -0,0 +1,232 @@
+import { Meta } from "@storybook/react";
+import { useEffect, useMemo, useState } from "react";
+import { checkboxFilter, multiFilter } from "src/components/Filters";
+import { GridDataRow } from "src/components/Table";
+import { collapseColumn, column, numericColumn, selectColumn } from "src/components/Table/utils/columns";
+import { simpleHeader } from "src/components/Table/utils/simpleHelpers";
+import { noop } from "src/utils";
+import { withBeamDecorator, withRouter, zeroTo } from "src/utils/sb";
+import { TestProjectLayout } from "../Layout.stories";
+import { GridTableLayout as GridTableLayoutComponent, useGridTableLayoutState } from "./GridTableLayout";
+
+export default {
+  component: GridTableLayoutComponent,
+  decorators: [withBeamDecorator, withRouter()],
+  parameters: { layout: "fullscreen" },
+} satisfies Meta;
+
+type Data = { name: string | undefined; value: number | undefined; status: string; priority: number };
+type HeaderRow = { kind: "header"; id: string; data: undefined };
+type ParentRow = { kind: "parent"; id: string; data: Data; children: GridDataRow<Row>[] };
+type DataRow = { kind: "data"; id: string; data: Data };
+type Row = HeaderRow | ParentRow | DataRow;
+
+export function GridTableLayout() {
+  const filterDefs = useMemo(() => getFilterDefs(), []);
+  const columns = useMemo(() => getColumns(), []);
+
+  const layoutState = useGridTableLayoutState({
+    persistedFilter: {
+      filterDefs,
+      storageKey: "grid-table-layout",
+    },
+    useSearch: "client",
+  });
+
+  return (
+    <TestProjectLayout>
+      <GridTableLayoutComponent
+        pageTitle="Grid Table Layout Example"
+        breadcrumb={[
+          { href: "/", label: "Home" },
+          { href: "/", label: "Sub Page" },
+        ]}
+        layoutState={layoutState}
+        tableProps={{
+          columns: [collapseColumn<Row>(), selectColumn<Row>(), ...columns],
+          rows: [simpleHeader, ...makeNestedRows(3)],
+          sorting: { on: "client", initial: [columns[1].id!, "ASC"] },
+        }}
+        primaryAction={{ label: "Primary Action", onClick: noop }}
+        secondaryAction={{ label: "Secondary Action", onClick: noop }}
+        tertiaryAction={{ label: "Tertiary Action", onClick: noop }}
+      />
+    </TestProjectLayout>
+  );
+}
+
+export function QueryTableLayout() {
+  const filterDefs = useMemo(() => getFilterDefs(), []);
+
+  const layoutState = useGridTableLayoutState({
+    persistedFilter: {
+      filterDefs,
+      storageKey: "grid-table-layout",
+    },
+    useSearch: "server",
+  });
+
+  // In this example, we set up a server-side search that uses the `searchString` from the layout state.
+  // in combination with the "QueryTable" behavior for loading/error states.
+  const query = useExampleQuery({ filter: { ...layoutState.filter, search: layoutState.searchString } });
```

Comment:
Does seem like the ordering of hooks is pretty nuanced here, i.e.:

* The query is going to need the filters as input ... so the `useLayoutState` must come first
* But any "row building" needs to use the backend query data ... so we'll never be able to put "rows" into `useLayoutState`

Which is fine for our current API that is passing the rows into `tableProps`, but just in general it seems like "someday TableLayoutState would like to know _something_ about the rows for some interesting behavior" would not be unheard of, and seems like that would be difficult to pull off.

Not because of your approach, which is working really well, but just because of the strict evaluation order of react hooks.

Reply:
Yeah for sure, this is one of those times where I miss the days of having a redux store that you can just share a central state that any level of the hierarchy can access/set and not deal with extra provider wrappers. The rules of hooks was also a real pain in this implementation as well since we can't conditionally call a hook, but we need the groupBy, filter & search behaviors to be opt-in.

Reply:
Yeah. Fwiw I like "non-global mobx stores that are just passed as props" for that sort of thing; it's what the GridTable headless state is managed with and has ended up being pretty clean.

Granted, still has to pass it as props, but ngl I kinda like knowing/seeing that dependency, vs even contexts which are "okay not truly global variables" but still "dependencies that you can't see".